### PR TITLE
Ingest building min_height OSM tag

### PR DESCRIPTION
**Pull request by arpanaj**
If a building does not start from ground level, then OSM min_height tag is used to indicate the height of bottom part of the building. Adding method to read this tag and set it in the Building entity.

### CODE REVIEW

File: src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package org.openstreetmap.atlas.tags;
+
+import org.openstreetmap.atlas.tags.annotations.Tag;
+import org.openstreetmap.atlas.tags.annotations.Tag.Validation;
+import org.openstreetmap.atlas.tags.annotations.TagKey;
+
+/**
+ * OSM min_height tag: https://taginfo.openstreetmap.org/keys/min_height#values
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java]: Nit: the values link can be removed, it's already available in the taginfo parameter below :)

File: src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+ *
+ * @author ajayaswal
+ */
+@Tag(value = Validation.LENGTH, taginfo = "https://taginfo.openstreetmap.org/keys/min_height#values", osm = "https://wiki.openstreetmap.org/wiki/Key:min_height")
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java]: Is `LENGTH` validator the correct one to use here? TagInfo shows a lot of negative values for this tag (-3, -6, etc.) If we want to consider these, maybe `DOUBLE` is the correct one to use.

Comment (MikeGost) [reply src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java]: Also, we may want to add a few unit test cases for this to reflect OSM data (negative values, 0, positives values, decimal values to make sure they're handled properly)

File: src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/buildings/ComplexBuilding.java [hunk 3]
```diff
@@ -58,6 +59,34 @@ protected ComplexBuilding(final AtlasEntity source)
...
+            minHeightTag = tags.get(MinHeightTag.KEY);
+            if (minHeightTag != null)
+            {
+                return Optional.of(HEIGHT_CONVERTER.convert(minHeightTag));
+            }
+        }
+        catch (final Exception e)
+        {
+            logger.warn("Invalid base height {} for building id {}", minHeightTag,
+                    getSource().getIdentifier());
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/buildings/ComplexBuilding.java]: Do we care about keeping the exception in the log?

File: src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/buildings/ComplexBuilding.java [hunk 4]
```diff
@@ -58,6 +59,34 @@ protected ComplexBuilding(final AtlasEntity source)
...
+    {
+        Map<String, String> tags = getSource().getTags();
+        String minHeightTag = tags.get(MinHeightTag.KEY);
+        try
+        {
+            if (minHeightTag != null)
+            {
+                return Optional.of(HEIGHT_CONVERTER.convert(minHeightTag));
+            }
+            tags = this.outlineSource.getTags();
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/buildings/ComplexBuilding.java]: Consider using `Validators.hasValuesFor(...)` here instead of grabbing allTags - this can be somewhat expensive.

File: src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java (hunk 2 above)
Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java]: Oh also it might be a problem using `Distance` which cannot be negative...

File: src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/buildings/ComplexBuilding.java [hunk 5]
```diff
@@ -58,6 +59,34 @@ protected ComplexBuilding(final AtlasEntity source)
...
+     */
+    public Optional<Distance> baseHeight()
+    {
+        Map<String, String> tags = getSource().getTags();
+        String minHeightTag = tags.get(MinHeightTag.KEY);
+        try
+        {
+            if (minHeightTag != null)
+            {
+                return Optional.of(HEIGHT_CONVERTER.convert(minHeightTag));
```
Comment (AlexHsieh22) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/buildings/ComplexBuilding.java]: Would it be possible to create a "getMinimumHeight" function that does:
```
try
{
    String minHeight = tags.get(MinHeightTag.KEY);

    if (minHeightTag != null)
    {
        return Optional.of(HEIGHT_CONVERTER.convert(minHeightTag));
    }
}
catch(...)
{
...
}
```

Then we could make 2 separate calls here and if the first call throw an exception, the second one can still be checked!

File: src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java (hunk 1 above)
Comment (AlexHsieh22) [reply src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java]: nit: Maybe a href for the link?

Comment (arpanaj) [reply src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java]: Removed the ink, keeping the one in the tag info parameter.

File: src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java (hunk 2 above)
Comment (arpanaj) [reply src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java]: Good point, will use a DoubleValidator and add more test cases. Will replace Distance with a new Height scalar.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/buildings/ComplexBuilding.java (hunk 5 above)
Comment (arpanaj) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/buildings/ComplexBuilding.java]: Will cleanup this code, looks like its more relevant for height, not for min_height.

File: src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    static Optional<Height> get(final Taggable taggable)
+    {
+        final Optional<String> tagValue = taggable.getTag(KEY);
+
+        if (tagValue.isPresent())
+        {
+            if (validator.isValid(tagValue.get()))
+            {
+                return Optional.of(new Height(Double.valueOf(tagValue.get())));
+            }
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java]: Nit: possible improvement could be combining if statements:
`if (tagValue.isPresent() && validator.isValid(tagValue.get())) {...}`

File: src/main/java/org/openstreetmap/atlas/utilities/scalars/Height.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
+package org.openstreetmap.atlas.utilities.scalars;
+
+import java.io.Serializable;
+
+/**
+ * Distance wrapper to denote height above or below ground
+ *
+ * @author ajayaswal
+ */
+
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/utilities/scalars/Height.java]: Nit: extra line

File: src/main/java/org/openstreetmap/atlas/utilities/scalars/Height.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    private final Distance distance;
+    private final boolean aboveGround;
+
+    public Height()
+    {
+        this.distance = Distance.ZERO;
+        this.aboveGround = true;
+    }
+
+    public Height(final Double rawHeight)
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/utilities/scalars/Height.java]: double instead of Double?

File: src/main/java/org/openstreetmap/atlas/utilities/scalars/Height.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    public double getAbsoluteHeightAsMeters()
+    {
+        return this.distance.asMeters();
+    }
+
+    public boolean isAboveGround()
+    {
+        return this.aboveGround;
+    }
+
```
Comment (AlexHsieh22) [inline src/main/java/org/openstreetmap/atlas/utilities/scalars/Height.java]: would a ```getHeighAsMeters()``` method be useful here?

File: src/main/java/org/openstreetmap/atlas/utilities/scalars/Height.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    @Override
+    public String toString()
+    {
+        return String.format("%.1f meters ", this.distance, this.aboveGround ? "" : "underground");
```
Comment (AlexHsieh22) [inline src/main/java/org/openstreetmap/atlas/utilities/scalars/Height.java]: are we missing a %s in the string format?

Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/utilities/scalars/Height.java]: Slim chance of this happening, but it looks like distance can be printed in kilometers in some cases, do we want to rely on the toString for distance to dictate units? Also, do we need a second string.format parameter for the "" or "underground" string?

File: src/main/java/org/openstreetmap/atlas/utilities/scalars/Height.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.io.Serializable;
+
+/**
+ * Distance wrapper to denote height above or below ground
+ *
+ * @author ajayaswal
+ */
+
+public class Height implements Serializable
```
Comment (AlexHsieh22) [inline src/main/java/org/openstreetmap/atlas/utilities/scalars/Height.java]: Thinking about this class a bit more, and after a brief discussion with @MikeGost , could we potentially use or extend the ```Altitude``` class?

File: src/main/java/org/openstreetmap/atlas/utilities/scalars/Height.java (hunk 8 above)
Comment (arpanaj) [reply src/main/java/org/openstreetmap/atlas/utilities/scalars/Height.java]: Removing this class, existing Altitude class works.

File: src/main/java/org/openstreetmap/atlas/utilities/scalars/Height.java (hunk 10 above)
Comment (arpanaj) [reply src/main/java/org/openstreetmap/atlas/utilities/scalars/Height.java]: Removing this class, existing Altitude class works.

File: src/main/java/org/openstreetmap/atlas/utilities/scalars/Height.java (hunk 11 above)
Comment (arpanaj) [reply src/main/java/org/openstreetmap/atlas/utilities/scalars/Height.java]: Thanks, Altitude class works well here, removing Height.

File: src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+import org.openstreetmap.atlas.tags.annotations.Tag.Validation;
+import org.openstreetmap.atlas.tags.annotations.TagKey;
+import org.openstreetmap.atlas.tags.annotations.validation.DoubleValidator;
+
+/**
+ * OSM min_height tag
+ *
+ * @author ajayaswal
+ */
+@Tag(value = Validation.LENGTH, taginfo = "https://taginfo.openstreetmap.org/keys/min_height#values", osm = "https://wiki.openstreetmap.org/wiki/Key:min_height")
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java]: Still think we should change this to Validation.DOUBLE - since `LengthValidator` would fail on a large number of values.

File: src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+    DoubleValidator validator = new DoubleValidator();
+
+    static Optional<Altitude> get(final Taggable taggable)
+    {
+        final Optional<String> tagValue = taggable.getTag(KEY);
+
+        if (tagValue.isPresent() && validator.isValid(tagValue.get()))
+        {
+            return Optional.of(Altitude.meters(Double.valueOf(tagValue.get())));
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java]: One minor thing to consider, there's a pattern of `validateAndExtract` found in other tag classes that contain numerical values (see `LayerTag` and `MaxSpeedForwardTag`). As an optional improvement, we should consider adding a `DoubleExtractor` and following the same pattern. This also works though!

File: src/test/java/org/openstreetmap/atlas/geography/atlas/items/complex/buildings/BuildingHeightTestCaseRule.java [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+ * Test fixture for the {@link BuildingHeightTestCase}
+ *
+ * @author ajayaswal
+ */
+public class BuildingHeightTestCaseRule extends CoreTestRule
+{
+    public static final long BUILDING_WITH_ZERO_MINHEIGHT_ID = 100000000L;
+    public static final long BUILDING_WITH_NEGATIVE_MINHEIGHT_ID = 200000000L;
+    public static final long BUILDING_WITH_BOTH_HEIGHT_ID = 300000000L;
+    public static final long BUILDING_WITH_NO_MINHEIGHTT_ID = 400000000L;
```
Comment (MikeGost) [inline src/test/java/org/openstreetmap/atlas/geography/atlas/items/complex/buildings/BuildingHeightTestCaseRule.java]: Nit: two `TT_ID` --> `T_ID`. Also, consider adding `_` between `MIN` and `HEIGHT` for readability.

File: src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java (hunk 12 above)
Comment (arpanaj) [reply src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java]: Agree, missed in the previous checkin, will fix now.

File: src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java (hunk 13 above)
Comment (arpanaj) [reply src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java]: Since the extraction code is very minimal, thinking of leaving it as is, instead of creating a new class.

Comment (MikeGost) [reply src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java]: Sounds good!

**Comment by arpanaj** [comment]
@matthieun , added a separate MinHeightTag class, all the tag value validations for MaxHeightTag do not apply to MinHeight.

File: src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+ *
+ * @author ajayaswal
+ */
+@Tag(value = Validation.DOUBLE, taginfo = "https://taginfo.openstreetmap.org/keys/min_height#values", osm = "https://wiki.openstreetmap.org/wiki/Key:min_height")
+public interface MinHeightTag
+{
+    @TagKey
+    String KEY = "min_height";
+
+    DoubleValidator validator = new DoubleValidator();
```
Comment (snb3300) [inline src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java]: Because this is getting used only in a static method below do you think it will be a good idea to have it as
``` private static final DoubleValidator doubleValidator = new DoubleValidator()``` ?

Comment (arpanaj) [reply src/main/java/org/openstreetmap/atlas/tags/MinHeightTag.java]: Its an interface, so can't do private.
